numeric_ranges_algorithms/numeric_ranges_algorithms.md

```diff
@@ -0,0 +1,867 @@
+
+### We don't propose `partial_sum`
+
+The `partial_sum` algorithm performs operations sequentially.  The existing ranges library does not have an equivalent algorithm with this left-to-right sequential behavior, nor do we propose such an algorithm.  For users who want this behavior, <a href="https://wg21.link/P2760R1">P2760R1</a> suggests a view instead of an algorithm.  <a href="https://wg21.link/P3351R2">P3351R2</a>, "`views::scan`," proposes this view.  P3351R2 is currently in SG9 (Ranges Study Group) review.
+
+Users of `partial_sum` who are not concerned about the order of operations can call `inclusive_scan` instead, which we propose here.  We considered adding a convenience wrapper for the same special case of an inclusive prefix plus-scan that `partial_sum` supports.  However, names like `partial_sum` or `prefix_sum` would obscure whether this is an inclusive or exclusive scan.  We think it's not a very convenient convenience wrapper if users have to look this up every time they use it.
+
+### We don't propose `reduce_first`
+
+Section 5.1 of <a href="https://wg21.link/P2760R1">P2760R1</a> asks whether there should be a `reduce_first` algorithm, analogous to `fold_left_first`, for binary operations that lack a natural identity element to serve as the initial value.  An example would be `min` on a range of `int` values, where callers would have no way to tell if `INT_MAX` represents a value in the range, or an arbitrary stand-in for the (nonexistent) identity element.  A `reduce_first` algorithm would use the first element in the range as the initial value.
+
+We do not propose `reduce_first` here.  We outline arguments against and for it, and then explain that a reasonable parallel generalization of `reduce_first` would actually be a different algorithm that can start with any element in the range.  We do not propose that generalization here, but would not oppose such a proposal.
+
+#### Arguments against `reduce_first`
+
+1. P3179R8 already proposes parallel ranges overloads of `min_element`, `max_element`, and `minmax_element`.  Minima and maxima are the main use cases that lack a natural identity element.
+
+2. A parallel `reduce_first` (or even the parallel analog described in (3)) would have potential performance issues.  Its computational loop would be less uniform.  This could introduce branches and data misalignment.  Fetching the first element of the entire range on all processors executing the reduction would reduce spatial locality.
+
+3. Existing parallel programming frameworks like OpenMP require an identity element or at least an initial value for the reduction operation, so they could not be used to implement `reduce_first`.  OpenMP permits custom reduction operations, so it would be a good candidate for implementing Standard parallel algorithms.  Hand-rolling `reduce_first` out of other OpenMP components would add implementation effort and could hinder compiler optimizations.

```

Comment:
And alas, these arguments equally apply to `reduce` as well.

> Existing parallel programming frameworks like OpenMP require an identity element or at least an initial value for the reduction operation

OpenMP specifically requires a way to initialize local accumulators. For the operations supported out of the box, it's the corresponding identity value (or the best approximation, as in the case of min/max). For user-defined operations, it has to be specified - either as a value or as a function. See https://www.openmp.org/spec-html/5.0/openmpsu107.html for details; some explanations: `omp_priv` there refers to a local accumulator, and `omp_orig` is essentially the initial value.

Reply:
I've fixed this and added OpenMP material in this section, under "Other parallel programming models."  Thanks!